### ISSUE TITLE
Failure in certificates : Generate CSR for the Server

### ISSUE THREAD

**Issue by yjml** [StreisandEffect/streisand#1080]
Getting an error in this task, deploying to a VM I'm running on my local network which is behind a dynamic IP (for which I'm inputting a domain name due to DDNS needs). 

`TASK [certificates : Generate CSR for the Server] **********************************************************************************************************************************
fatal: [scrubbed.domain.net]: FAILED! => {"changed": true, "cmd": ["openssl", "req", "-batch", "-extensions", "server", "-new", "-key", "server.key", "-out", "server.csr", "-config", "/etc/ocserv/openssl.cnf"], "delta": "0:00:00.031325", "end": "2017-12-04 16:54:11.035660", "msg": "non-zero return code", "rc": 1, "start": "2017-12-04 16:54:11.004335", "stderr": "Error Loading extension section server\n140363043792536:error:220A4076:X509 V3 routines:a2i_GENERAL_NAME:bad ip address:v3_alt.c:476:value=scrubbed.domain.net\n140363043792536:error:22098080:X509 V3 routines:X509V3_EXT_nconf:error in extension:v3_conf.c:95:name=subjectAltName, value=@alt_names", "stderr_lines": ["Error Loading extension section server", "140363043792536:error:220A4076:X509 V3 routines:a2i_GENERAL_NAME:bad ip address:v3_alt.c:476:value=scrubbed.domain.net", "140363043792536:error:22098080:X509 V3 routines:X509V3_EXT_nconf:error in extension:v3_conf.c:95:name=subjectAltName, value=@alt_names"], "stdout": "", "stdout_lines": []}`

### Ansible Information

* Ansible version: 2.4.2.0
* Ansible system: FreeBSD
* Host OS: FreeBSD
* Host OS version:  FreeBSD 11.0-STABLE #0 r321665+25fe8ba8d06(freenas/11.0-stable): Mon Sep 25 06:24:11 UTC 2017     root@gauntlet:/freenas-11-releng/freenas/_BE/objs/freenas-11-releng/freenas/_BE/os/sys/FreeNAS.amd64
* Python interpreter: python
* Python version: 2.7.13

### Streisand Information

* Streisand Git revision: 975f54f7a72a2daaa510037dc41143fb12a3f9f9
* Streisand Git clone has untracked changes: no
* Genesis role: existing-server
* Custom SSH key: False

### Enabled Roles

* Shadowsocks enabled:  True
* Wireguard enabled: True
* OpenVPN enabled: True
* stunnel enabled: True
* Tor enabled: False
* Openconnect enabled: True
* TinyProxy enabled: False
* SSH forward user enabled: True
* L2TP enabled: True
* Configured number of VPN clients: 5

**Comment by cpu** [comment]
Hi @yjml,

> Getting an error in this task, deploying to a VM I'm running on my local network which is behind a dynamic IP (for which I'm inputting a domain name due to DDNS needs).

I believe this is the problem. Streisand expects an IP address and you gave it a domain name. The domain name you provide in place of the IP address is [eventually being written to an IP type Subject Alternate Name](https://github.com/StreisandEffect/streisand/blob/975f54f7a72a2daaa510037dc41143fb12a3f9f9/playbooks/roles/certificates/templates/openssl.cnf.j2#L53) in the OCServ certificate generation. OpenSSL is erroring because the IP is not valid.

> Error Loading extension section server", "140363043792536:error:220A4076:X509 V3 routines:a2i_GENERAL_NAME:bad ip address:v3_alt.c:476:value=scrubbed.domain.net

Unfortunately Streisand doesn't support the DDNS deployment you're trying to use it for out of box. We don't have the resources to prioritize this particular usage model but if you are interested in trying to submit patches I'd be happy to review them.

**Comment by yjml** [comment]
Good to know. I know this used work about half a year back (looking at dates of generated docs - 5/28/17), but I accidentally trashed the setup recently and need to recreate. Looking at that commit, the IP requirement was added with the refactor in October, so that makes sense.

I'm not familiar at all with whats going on in the background but I'll play around with things (aka do some line deletion) and see if things break further.

**Comment by yjml** [comment]
Deletion of lines no good (as expected). 

Backing out to the commit [prior](https://github.com/StreisandEffect/streisand/commit/b611142689284bdc4e2fd9994651b68cc3cf7358) to the OpenSSL refactor and got things running.

**Comment by yjml** [comment]
And removing the IP related lines allows a successful install, though I have no idea if this would cause any problems. Won't be able to test OpenVPN until tomorrow.

```diff
diff --git a/playbooks/roles/certificates/templates/openssl.cnf.j2 b/playbooks/roles/certificates/templates/openssl.cnf.j2
index e81bea6..9424288 100644
--- a/playbooks/roles/certificates/templates/openssl.cnf.j2
+++ b/playbooks/roles/certificates/templates/openssl.cnf.j2
@@ -43,15 +43,6 @@ emailAddress = optional

 [ req ]
 distinguished_name = req_distinguished_name
-req_extensions     = req_ext
-
-[ req_ext ]
-subjectAltName = @alt_names
-
-[ alt_names ]
-{% for item in tls_sans %}
-IP.{{ loop.index }} = {{ item }}
-{% endfor %}

 [ req_distinguished_name ]
 countryName = Country Name (2 letter code)
@@ -82,7 +73,6 @@ subjectKeyIdentifier=hash
 authorityKeyIdentifier=keyid,issuer:always
 extendedKeyUsage=serverAuth
 keyUsage = digitalSignature, keyEncipherment
-subjectAltName = @alt_names

 [ client ]
 basicConstraints=CA:FALSE
```

**Comment by cpu** [comment]
Hi @yjml,

> And removing the IP related lines allows a successful install, though I have no idea if this would cause any problems. 

I suspect based on the diff above that your certificates won't have the correct subject and will likely be rejected by OpenVPN.

I'm going to close this issue for now since we've identified the root cause and there isn't any unexpected errors to address, only potential future work. Thanks!

**Comment by craSH** [comment]
Just commenting as I came across this, and may (without taking too much responsibility!) put some effort in to seeing if I can make something work for dynamic DNS situations.

Placing the IP address in the `alt_names` in this case would be difficult to deal with should your IP change in the future, but it wouldn't necessarily break functionality if you were accessing the service by the DNS name which is updated via whatever means.

I've ran this in a half-way there way by doing a DNS query to find out the current IP address of my host, but that still uses the IP in configuration files which becomes useless once it changes in the future.

E.g. in _site.yml_, after ensuring the Python package _dnspython_ is installed:

```
streisand_ipv4_address: "{{ lookup('dig', 'some.dynamic.dns.host.example.com')}}"
```

Perhaps an additional variable could be introduced like `streisand_public_address` that could be used in configuration templates where `streisand_ipv4_address` is currently used, and by default it can just be the value of `streisand_ipv4_address` itself. Users could manually specify it to be a DDNS hostname in _site.yml_, with a stated caveat of `alt_names` and anything similar to that (that I'm not aware of yet) perhaps not being 100% ideal.
